src/provider/dynamodb.ts

```diff
@@ -0,0 +1,198 @@
+    private dynamoDB: DynamoDB;
+    private initialized = false;
+
+    constructor(awsConfig: AWSConfig) {
+        AWS.config.update(awsConfig);
+
+        this.documentClient = new DynamoDB.DocumentClient();
+        this.dynamoDB = new AWS.DynamoDB();
+    }
+
+    public async addEvent(stream: Stream, data: any) {
+        await this.ensureTables();
+
+        this.addAggregation(stream);
+        const commitTimestamp = Date.now();
+        const item = {
+            aggregation_streamid: `${this.getKey(stream)}`,
+            commitTimestamp: commitTimestamp,
+            payload: data,
+            stream: stream
+        };
+        const param = {
+            Item: item,
+            TableName: 'events',
+        };
+
+        await this.documentClient.put(param, (error, _) => {
+            if (error) {
+                throw new Error(error.message);
+            }
+        });
+
+        return {
+            commitTimestamp: commitTimestamp,
+            payload: data,
+        } as Event;
+    }
+
+
+    public async getEvents(stream: Stream, offset: number = 0, limit: number = -1): Promise<Array<Event>> {
+        await this.ensureTables();
+        const params = {
+            ConsistentRead: true,
+            ExpressionAttributeValues: {
+                ':a': this.getKey(stream)
+            },
+            KeyConditionExpression: 'aggregation_streamid = :a',
+            ScanIndexForward: false,
+            TableName: 'events',
+        };
+
+        const items: ItemList = (await this.documentClient.query(params).promise()).Items;
+
+        return items.map(data => {
+            return {
+                commitTimestamp: data.commitTimestamp,
+                payload: data.payload,
+            } as Event;
+        });
+    }
+
+    public async getAggregations(offset: number = 0, limit: number = -1): Promise<Array<string>> {
+        await this.ensureTables();
+        const params = {
+            TableName: 'aggregations',
+        };
+
+        const items = await (await this.documentClient.scan(params).promise());
+
+        return items.Items.map(data => data.aggregation);
+    }
+
+    public async getStreams(aggregation: string, offset: number = 0, limit: number = -1): Promise<Array<string>> {
+        await this.ensureTables();
+        const params = {
+            ConsistentRead: true,
+            ExpressionAttributeValues: {
+                ':a': aggregation
+            },
+            KeyConditionExpression: 'aggregation = :a',
+            ScanIndexForward: false,
+            TableName: 'aggregations',
+        };
+
+        const items = await (await this.documentClient.query(params).promise()).Items;
+
+        return items.map(data => data.stream);
+    }
+
+    private async addAggregation(stream: Stream) {
+        const param = {
+            Item: {
+                aggregation: stream.aggregation,
+                stream: stream.id
+            },
+            TableName: 'aggregations',
```

Comment:
Same case as mentioned before.

Reply:
I didn't understand this suggestion, shall we talk?